exercises/yacht/examples/success-standard/src/Yacht.hs

```diff
@@ -0,0 +1,69 @@
+module Yacht (yacht, Category(..)) where
+
+import Data.List
+
+data Category = Ones
+              | Twos
+              | Threes
+              | Fours
+              | Fives
+              | Sixes
+              | FullHouse
+              | FourOfAKind
+              | LittleStraight
+              | BigStraight
+              | Choice
+              | Yacht
+
+yacht :: Category -> [Int] -> Int
+yacht _ []   = error "You lose!"
+yacht Ones   dice = simpleScore 1 dice
+yacht Twos   dice = simpleScore 2 dice
+yacht Threes dice = simpleScore 3 dice
+yacht Fours  dice = simpleScore 4 dice
+yacht Fives  dice = simpleScore 5 dice
+yacht Sixes  dice = simpleScore 6 dice
+
+yacht FullHouse dice
+  | s == 2 && length g == 2 && (length . head') gs == 3 = sum dice
+  | otherwise = 0
+  where
+    s = (length . gsortBySize) dice
+    (g:gs) = gsortBySize dice
+
+yacht FourOfAKind dice
+  | length lg == 4 = sum lg
+  | length lg == 5 = 4 *  (head dice)
+  | otherwise = 0
+  where
+    lg = largestGroup dice
+
+yacht LittleStraight dice
+  |  isSucc dice && maximum dice == 5 = 30
+  | otherwise = 0
+
+yacht BigStraight dice
+  | isSucc dice && maximum dice == 6 = 30
+  | otherwise = 0
+
+yacht Choice dice = sum dice
+
+yacht Yacht (d:ice)
+  | all (== d) ice = 50
+  | otherwise      = 0
+
+simpleScore :: Int -> [Int] -> Int
+simpleScore x dice = (sum . filter (== x)) dice
+
+gsortBySize :: ((Ord a), Eq a) => [a] -> [[a]]
+gsortBySize = sortOn (length) . group . sort
+
+largestGroup :: ((Ord a), Eq a) => [a] -> [a]
+largestGroup = head . reverse . gsortBySize
+
+isSucc :: [Int] -> Bool
+isSucc = all (\(x,y) -> succ x == y) . pairwise . sort
+  where pairwise = zip <*> tail
+
+head' [] = []
```

Comment:
Missing signature. I'm thinking maybe you could do something else here, but I can't think of what at this moment.

Reply:
That is exactly why I left the signature off of this.  I was hoping we might figure out how to avoid it altogether.

Reply:
You could avoid having `head'` by doing concrete pattern matching:

```haskell
yacht FullHouse dice = case sortDiceByGroupSize dice of
  [[_, _], [_, _, _]] -> 50
  _ -> 0
```

I once came to the conclusion that direct pattern matching was always preferrable (in a context of pattern matching much larger ASTs in compilers), and I remember a professor saying that direct pattern matching is less abstract: If you decide to change the representation, you end up having to replace code in more places than if you always address your internal representation abstractly.

You may for example want to change the internal representation so that the largest dice group is ordered first, or so that it's a `[(Die, Count)]` rather than a `[[Die]]`. So if you want to avoid direct pattern matching and partial functions, and use abstract helper functions, maybe something like:

```haskell
yacht FullHouse dice
  | (length <$> largestDiceGroup dice) == Just 3 &&
    (length <$> nextLargestDiceGroup dice) == Just 2
    = sum dice
  | otherwise = 0

largestDice :: [Int] -> Maybe [Int]
largestDice = lastMay . sortDiceByGroupSize

nextLargestDice :: [Int] -> Maybe [Int]
nextLargestDice dice = (\xs -> initMay xs >>= lastMay) . sortDiceByGroupSize
```

The functions `lastMay` and `initMay` are from the package [`safe`](https://hackage.haskell.org/package/safe/docs/Safe.html). In this case, I'd probably just go with pattern matching because with the scope of the exercise, I don't feel like creating a layer of abstraction between `yacht` and the way I represent dice groups.

Reply:
Thanks for the advice.  I like `fmap` there 😁